components/brave_wallet/browser/eth_abi_decoder.cc

```diff
@@ -0,0 +1,299 @@
+// calldata at the specified offset.
+absl::optional<uint256_t> GetUint256FromData(const std::vector<uint8_t>& input,
+                                             size_t offset) {
+  auto value = GetArgFromData(input, offset);
+  if (!value)
+    return absl::nullopt;
+
+  uint256_t arg;
+  if (!HexValueToUint256("0x" + *value, &arg))
+    return absl::nullopt;
+  return arg;
+}
+
+// GetUint256FromData extracts a 32-byte wide uint256 hex value from the
+// calldata at the specified offset.
+//
+// The parsed uint256 value is serialized as a compact hex string (without)
+// leading 0s, and prefixed by "0x".
+absl::optional<std::string> GetUint256HexFromData(
+    const std::vector<uint8_t>& input,
+    size_t offset) {
+  auto arg_uint256 = GetUint256FromData(input, offset);
+  if (!arg_uint256)
+    return absl::nullopt;
+
+  return Uint256ValueToHex(*arg_uint256);
+}
+
+// GetSizeFromData extracts a 32-byte wide size_t value from the calldata
+// at the specified offset.
+//
+// Using this function to extract an integer outside the range of size_t is
+// considered an error. Ideal candidates are calldata tail references, length
+// of dynamic types, etc.
+absl::optional<size_t> GetSizeFromData(const std::vector<uint8_t>& input,
+                                       size_t offset) {
+  auto value = GetUint256FromData(input, offset);
+  if (!value)
+    return absl::nullopt;
+
+  // Since we use value as an array index, we need to cast the type to
+  // something that can be used as an index, viz. size_t. To prevent runtime
+  // errors, we make sure the value is within safe limits of size_t.
+  if (*value > std::numeric_limits<size_t>::max())
+    return absl::nullopt;
+
+  return static_cast<size_t>(*value);
+}
+
+// GetBoolFromData extracts a 32-byte wide boolean value from the
+// calldata at the specified offset.
+//
+// The parsed bool value is serialized as "true" or "false" strings.
+absl::optional<std::string> GetBoolFromData(const std::vector<uint8_t>& input,
+                                            size_t offset) {
+  auto value = GetUint256FromData(input, offset);
+  if (!value)
+    return absl::nullopt;
+
+  if (value == 0)
+    return "false";
+  else if (value == 1)
+    return "true";
+
+  return absl::nullopt;
+}
+
+// GetBytesHexFromData extracts a bytes value from the calldata at the
+// specified offset using head-tail encoding mechanism. bytes are packed
+// tightly in chunks of 32 bytes, with the first 32 bytes encoding the length,
+// followed by the actual content.
+//
+// The parsed bytearray is serialized as a hex string prefixed by "0x".
+absl::optional<std::string> GetBytesHexFromData(
+    const std::vector<uint8_t>& input,
+    size_t offset) {
+  auto pointer = GetSizeFromData(input, offset);
+  if (!pointer)
+    return absl::nullopt;
+
+  auto bytes_len = GetSizeFromData(input, *pointer);
+  if (!bytes_len)
+    return absl::nullopt;
+
+  if (input.size() < *pointer + 32 + *bytes_len)
```

Comment:
In theory `*pointer + 32 + *bytes_len` could overflow to smth less than `input.size()`

Reply:
Now casting the `size_t` values to a `uint256_t`, so it's a bit more defensive, but as discussed internally we need to have hard upper limits to harden the parser even more.